packages/vincent-tools/vincent-tool-sdk/src/lib/types.ts

```diff
@@ -0,0 +1,245 @@
+        ? PolicyResultDeny<z.infer<PrecheckDenyResult>>
+        : PolicyResultDenyNoResult)
+  >;
+
+  commitParamsSchema: NonNullable<CommitParams>;
+  commitAllowResultSchema?: CommitAllowResult;
+  commitDenyResultSchema?: CommitDenyResult;
+  commit: (
+    args: CommitParams extends z.ZodType ? z.infer<CommitParams> : never,
+  ) => Promise<
+    | (CommitAllowResult extends z.ZodType
+        ? PolicyResultAllow<z.infer<CommitAllowResult>>
+        : PolicyResultAllowNoResult)
+    | (CommitDenyResult extends z.ZodType
+        ? PolicyResultDeny<z.infer<CommitDenyResult>>
+        : PolicyResultDenyNoResult)
+  >;
+}
+
+// Union type for all policy definitions
+// Do not be alarmed by the `any` types here -- the actual policies are still constrained by the types that make up the union
+export type VincentPolicyDef =
+  | BasicPolicyDef<any, any, any, any>
+  | PolicyWithPrecheck<any, any, any, any, any, any>
+  | PolicyWithCommit<any, any, any, any, any, any, any>
+  | PolicyWithPrecheckAndCommit<any, any, any, any, any, any, any, any, any>;
+
+// Helper type to determine if a policy has commit
+export type HasCommit<P> = P extends { commit: infer CommitFn }
+  ? CommitFn extends Function
+    ? true
+    : false
+  : false;
+
+// Tool supported policy with proper typing
+export type VincentToolSupportedPolicy<
+  ToolParamsSchema extends z.ZodType,
+  PolicyDefType extends VincentPolicyDef,
+> = {
+  policyDef: PolicyDefType;
+  toolParameterMappings: Partial<{
+    [K in keyof z.infer<ToolParamsSchema>]: keyof z.infer<
+      PolicyDefType['toolParamsSchema']
+    >;
+  }>;
+};
+
+// Policy evaluation results with proper typing for commit
+export type VincentPolicyEvaluationResults<
+  Policies extends Record<string, { policyDef: VincentPolicyDef }>,
+> = {
+  evaluatedPolicies: Array<keyof Policies>;
+  allowPolicyResults: {
+    [PolicyKey in keyof Policies]?: {
+      result: PolicyResultAllow<any>;
+    } & (HasCommit<Policies[PolicyKey]['policyDef']> extends true
+      ? {
+          commit: Extract<
+            Policies[PolicyKey]['policyDef'],
+            { commit: Function }
+          >['commit'];
+        }
+      : {});
+  };
+} & (
+  | { allow: true; denyPolicyResult?: never }
+  | {
+      allow: false;
+      denyPolicyResult: {
+        result: PolicyResultDeny<any>;
+        ipfsCid: keyof Policies;
+      };
+    }
+);
+
+// Tool definition
+export interface VincentToolDef<
+  ToolParamsSchema extends z.ZodType,
+  Policies extends Record<
+    string,
+    VincentToolSupportedPolicy<ToolParamsSchema, VincentPolicyDef>
+  >,
+> {
+  toolParamsSchema: ToolParamsSchema;
+  supportedPolicies: Policies;
+  precheck: (
+    params: z.infer<ToolParamsSchema>,
+    policyEvaluationResults: VincentPolicyEvaluationResults<Policies>,
+  ) => Promise<unknown>;
+  execute: (
+    params: z.infer<ToolParamsSchema>,
+    policyEvaluationResults: VincentPolicyEvaluationResults<Policies>,
+  ) => Promise<unknown>;
+}
```

Comment:
Don't we want the return types to have some structure? e.g.

```tsc
{ success: true, result: unknown }

{ success: false, reason: string }
```

or something similar?

Reply:
:+1: Pushed up proper support for schema-driven return values for precheck and execute on the VincentTool -- sorry for the oversight. I also added the same context pattern to these methods and fleshed out the playground into a set of files that will actually fail to compile if we break any of the inference rules.